vita3k/regmgr/src/regmgr.cpp

```diff
@@ -0,0 +1,361 @@
+// Vita3K emulator project
+// Copyright (C) 2023 Vita3K team
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation; either version 2 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+#include <regex>
+#include <regmgr/functions.h>
+
+#include <util/log.h>
+
+namespace regmgr {
+
+const std::vector<unsigned char> xorKey = { 0x89, 0xFA, 0x95, 0x48, 0xCB, 0x6D, 0x77, 0x9D, 0xA2, 0x25, 0x34, 0xFD, 0xA9, 0x35, 0x59, 0x6E };
+
+static std::string decryptRegistryFile(const fs::path reg_path) {
+    std::ifstream file(reg_path.string(), std::ios::binary);
+    if (!file.is_open()) {
+        LOG_WARN("Error while opening file: {}, install firmware for solve this!", reg_path.string());
+        return {};
+    }
+
+    // Read the data from the file encrypted
+    std::vector<uint8_t> encryptedData((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+
+    // Check if file is empty
+    if (encryptedData.empty()) {
+        LOG_DEBUG("File is empty: {}", reg_path.string());
+        return {};
+    }
+
+    // Remove the header of the file (138 bytes)
+    encryptedData.erase(encryptedData.begin(), encryptedData.begin() + 138);
+
+    // Decrypt the data with the xor key
+    for (size_t i = 0; i < encryptedData.size(); i++) {
+        encryptedData[i] ^= xorKey[i & 0xF];
+    }
+
+    // Convert the data to a string
+    std::string res;
+    for (const auto &byte : encryptedData) {
+        res += byte;
+    }
+
+    return res;
+}
+
+static std::unordered_map<std::string, std::vector<RegValue>> reg_template;
+
+void init_reg_template(RegMgrState &regmgr, const std::string &reg) {
+    std::unordered_map<int, std::string> reg_map;
+
+    std::istringstream iss(reg);
+    std::string line;
+    while (std::getline(iss, line)) {
+        if (line.empty())
+            continue;
+
+        // Found the base register
+        if (line.find("[BASE") != std::string::npos) {
+            // Register the names with their numbers
+            while (std::getline(iss, line) && (line != "[REG-BAS")) {
+                if (line.empty())
+                    continue;
+
+                std::size_t pos = line.find('=');
```

Comment:
```suggestion
                size_t pos = line.find('=');
```
and same for the others

Reply:
ah yes

Reply:
done